spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageResource.java

```diff
@@ -0,0 +1,332 @@
+	}
+
+	/**
+	 * Creates the blob that this {@link GoogleStorageResource} represents in Google Cloud
+	 * Storage.
+	 * @return the created blob object
+	 * @throws StorageException if any errors during blob creation arise,
+	 * such as if the blob already exists
+	 * @throws IllegalStateException if the resource reference is to a bucket, and not a blob.
+	 */
+	public Blob createBlob() throws StorageException {
+		return this.storage.create(BlobInfo.newBuilder(getBlobId()).build());
+	}
+
+	/**
+	 * Creates the bucket that this resource references in Google Cloud Storage.
+	 * @return the {@link Bucket} object for the bucket
+	 * @throws StorageException if any errors during bucket creation arise,
+	 * such as if the bucket already exists
+	 */
+	public Bucket createBucket() {
+		return this.storage.create(BucketInfo.newBuilder(this.bucketName).build());
+	}
+
+	/**
+	 * Returns the {@link Bucket} associated with the resource.
+	 * @return the bucket if it exists, or null otherwise
+	 */
+	public Bucket getBucket() {
+		return this.storage.get(this.bucketName);
+	}
+
+	/**
+	 * Checks for the existence of the {@link Bucket} associated with the resource.
+	 * @return true if the bucket exists
+	 */
+	public boolean bucketExists() {
+		return getBucket() != null;
+	}
+
+	private Blob throwExceptionForNullBlob(Blob blob) throws IOException {
+		if (blob == null) {
+			throw new FileNotFoundException("The blob was not found: " + this.location);
+		}
+		return blob;
+	}
+
+	@Override
+	public File getFile() throws IOException {
+		throw new UnsupportedOperationException(
+				getDescription() + " cannot be resolved to absolute file path");
+	}
+
+	@Override
+	public long contentLength() throws IOException {
+		return throwExceptionForNullBlob(getBlob()).getSize();
+	}
+
+	@Override
+	public long lastModified() throws IOException {
+		return throwExceptionForNullBlob(getBlob()).getUpdateTime();
+	}
+
+	/**
+	 * Creates a file in the same bucket as this one. It inherits {@code autoCreateFiles}
+	 * from this object.
+	 *
+	 * <p>The file is created if it didn't exist in Google Cloud Storage.
+	 * @param relativePath the URL to a Google Cloud Storage file
+	 * @return the {@link GoogleStorageResource} object for the created file
+	 * @throws IOException
+	 */
+	@Override
+	public GoogleStorageResource createRelative(String relativePath) throws IOException {
+		GoogleStorageResource blobResource = new GoogleStorageResource(
+				this.storage, this.location.resolve(relativePath).toString(), this.autoCreateFiles);
+
+		if (!blobResource.exists()) {
+			blobResource.createBlob();
+		}
+
+		return blobResource;
+	}
+
+	@Override
+	public String getFilename() {
+		return isBucket() ? this.bucketName : this.blobName;
+	}
+
+	@Override
+	public String getDescription() {
+		return this.location.toString();
+	}
+
+	@Override
+	public InputStream getInputStream() throws IOException {
+		return Channels.newInputStream(throwExceptionForNullBlob(getBlob()).reader());
```

Comment:
Should we explicitly check if this resource is a bucket rather than relying on the vague exception thrown by `throwExceptionForNullBlob()`?